returnn/util/file_cache.py

```diff
@@ -324,19 +349,27 @@ def _copy_file_if_needed(self, src_filename: str, dst_filename: str):
         dst_dir = os.path.dirname(dst_filename)
         os.makedirs(dst_dir, exist_ok=True)
 
+        lock_dir, lock_file = self._get_lock_filename(dst_filename)
+        info_file_name = self._get_info_filename(dst_filename)
+
         # Copy the file, while holding a lock. See comment on lock_timeout above.
         with LockFile(
-            directory=dst_dir, name=os.path.basename(dst_filename) + ".lock", lock_timeout=self._lock_timeout
+            directory=lock_dir, name=lock_file, lock_timeout=self._lock_timeout
         ) as lock, self._touch_files_thread.files_added_context(lock.lockfile):
             # Maybe it was copied in the meantime, while waiting for the lock.
             if self._check_existing_copied_file_maybe_cleanup(src_filename, dst_filename):
                 print(f"FileCache: using existing file {dst_filename}")
+                # Update mtime while holding lock, to synchronize with any concurrent cleanup.
+                #
+                # See https://github.com/rwth-i6/returnn/issues/1675 for discussion.
+                os.utime(dst_filename, None)
+                os.utime(info_file_name, None)
                 return
 
             print(f"FileCache: Copy file {src_filename} to cache")
 
             # Make sure we have enough disk space, st_size +1 due to _copy_with_prealloc
-            self.cleanup(need_at_least_free_space_size=os.stat(src_filename).st_size + 1)
+            self.cleanup(lock_held_for=lock, need_at_least_free_space_size=os.stat(src_filename).st_size + 1)
```

Comment:
I don't understand, why do you pass `lock_held_for`? Why is it specifically relevant what lock this current proc is holding, but irrelevant what other procs are holding? I think in `cleanup`, all locks should be treated the same. The lock of the current proc should not get any special treatment.

Reply:
Ah, I realize now we will never be holding a lock for the file that we are trying to fetch here. If it existed, we would have used it. So agree, this code can go away. Fixing...